Upgrade to Calckey 13.1.1

lapineige opened the pull request:
v13.0.6 will be released soon.
It introduces some changes in the install process : #10 

~Codeberg is not working right know so I can't download the source archive and sha256sum it, hence this is still a draft.~ Downloading the archive is working again.

## PR Status
- [x] @oufmilo please merge master to testing :) #13
- [x] Code finished and ready to be reviewed/tested
- [x] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

lapineige committed:
yarn : switch to pnpm

Necessary for v13.0.6

lapineige committed:
yarn : switch to pnpm

lapineige committed:
update to v13.0.6-rc

Release candidate, but it's a way to try #10 changes.

lapineige commented:
!testme

lapineige committed:
typo + reordering + remove yarn

lapineige committed:
proper corepack rights

lapineige commented:
!testme

lapineige commented:
!testme

lapineige commented:
!testme

lapineige commented:
On this line: https://github.com/lapineige/calckey_ynh/blob/master/scripts/install#L137
```
sudo -u calckey env PATH=/opt/node_n/n/versions/node/19/bin:/opt/node_n/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin pnpm i
Usage Error: This project is configured to use yarn
```

I'm lost with this one… The script does not even refers to yarn…
@ThatOneCalculator : I think I need your help it, to do the conversion to pnpm… #10 

Also I don't think we use `yarn workspace`, am I mistaken ? Should we use it ?

lapineige commented:
Ok, I understood something ([source](https://pnpm.io/cli/install)): `pnpm i` is an alias for `pnpm install`.
I hate those very inexplicit naming :(

So I probably messed up the command order.

edit : I didn't have a better luck…

lapineige committed:
remove double + reordering

lapineige commented:
!testme

lapineige commented:
Reading [this documentation](https://yunohost.org/en/packaging_apps_helpers#ynh-use-nodejs), I'm wondering… @YunoHost-Apps/apps-group, is it possible that yunohost helpers are not made to work with `pnpm` as a replacement of `npm` ? 🤔

lapineige committed:
trying something…

lapineige commented:
!testme

lapineige committed:
pnpm: What if we remove yunohost helper…

lapineige commented:
!testme

lapineige commented:
Trying without Yunohost helpers

oufmilo commented:
pnpm is for the next update.. that's why it doesn't work

lapineige commented:
Oh it's not for 13.0.6 (release candidate) ? Planned for v14 ?

ThatOneCalculator commented:
v14 won't be releasing for 6+ months...

The next update with pnpm will either be v13.0.7 or v13.1.0

yalh76 commented:
> Reading [this documentation](https://yunohost.org/en/packaging_apps_helpers#ynh-use-nodejs), I'm wondering… @YunoHost-Apps/apps-group, is it possible that yunohost helpers are not made to work with `pnpm` as a replacement of `npm` ? 🤔

nope, it's only made to work with npm, you should look at the code behind the helper : https://github.com/YunoHost/yunohost/blob/394907ff0df218ec926f43c2e2ba006c68cc3f8a/helpers/nodejs#L74

lapineige commented:
Ok thanks.
Any change it gets adapted in the future ?

Or should we manage pnpm ourselves for this package ?

yalh76 commented:
> Ok thanks. Any change it gets adapted in the future ?
> 
> Or should we manage pnpm ourselves for this package ?

I don't know pnpm, but as far i can see, yunohost nodejs helper use https://github.com/tj/n,  , and n doesn't manage pnpm

lapineige commented:
@oufmilo [found a way](https://github.com/YunoHost-Apps/misskey_ynh/pull/145/files) in Misskey app, I'm replicating it here.

lapineige committed:
Another pnpm try…

From https://github.com/YunoHost-Apps/misskey_ynh/pull/145/files

lapineige committed:
Another pnpm try

lapineige commented:
!testme

lapineige committed:
Temporary version number

https://github.com/YunoHost-Apps/calckey_ynh/pull/12#issuecomment-1383218046

oufmilo commented:
> @oufmilo [found a way](https://github.com/YunoHost-Apps/misskey_ynh/pull/145/files) in Misskey app, I'm replicating it here.

Yes, I think it's a good way but for the moment calckey doesn't have release the version for pnpm, we should be patient.

lapineige commented:
Yeah I forgot that before launching the CI 😅

lapineige committed:
v13.1.0 - pnpm

Inspired by this : https://codeberg.org/calckey/calckey/releases

lapineige commented:
Failure : https://ci-apps-dev.yunohost.org/ci/job/5382

lapineige committed:
v13.1.0

lapineige commented:
!testme

lapineige committed:
remove  --frozen-lockfile

lapineige commented:
!testme

lapineige commented:
> @oufmilo please merge master to testing :) https://github.com/YunoHost-Apps/calckey_ynh/pull/13

ping @oufmilo :)

lapineige commented:
!testme

lapineige commented:
Install works ! :tada:

oufmilo commented:
[![Test Badge](https://img.shields.io/endpoint?url=https://yunorunner.travelpandas.eu/ci/api/job/9/badge)](https://yunorunner.travelpandas.eu/ci/job/9)
[![](https://yunorunner.travelpandas.eu/ci/summary/9.png)](https://yunorunner.travelpandas.eu/ci/job/9)

lapineige commented:
Wow that's way faster on your side :-o

(CI still running, there were a lock file issue)

lapineige committed:
pnpm

lapineige commented:
Restore fails when starting systemd service :

> npm[14220]: > calckey@13.1.0 start
npm[14220]: > pnpm --filter backend run start
npm[14237]: sh: 1: pnpm: not found

Should it run corepack enable after restoration ? :thinking:

lapineige committed:
pnpm migration + yarn dependency not needed ?

Yarn not installed on install, why on upgrade ?

And https://codeberg.org/calckey/calckey/releases/tag/v13.1.0

lapineige commented:
!testme

lapineige committed:
Update restore

Trying to remove `ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  backend@ migrate: `typeorm migration:run -d ormconfig.js` restoration error.

lapineige committed:
reordering +  remove frozen-lockfile

lapineige commented:
!testme

lapineige commented:
Hey @ThatOneCalculator the sha256sum of the archive changed, do you know why ?

lapineige committed:
new sha256sum

lapineige commented:
!testme

ThatOneCalculator commented:
> Hey @ThatOneCalculator the sha256sum of the archive changed, do you know why ?

Of which archive? We did reupload our docker image due to an issue.

lapineige commented:
``` 
779179 ERROR Could not restore calckey: An error occured inside the app restore script
779180 INFO The operation 'Restore 'calckey' from a backup archive' could not be completed. Please share the full log of this operation using the command 'yunohost log share 20230202-231805-backup_restore_app-calckey' to get help
779288 WARNING Here's an extract of the logs before the crash. It might help debugging the error:
779289 INFO DEBUG -   hint: undefined,
779290 INFO DEBUG -   position: undefined,
779290 INFO DEBUG -   internalPosition: undefined,
779291 INFO DEBUG -   internalQuery: undefined,
779291 INFO DEBUG -   where: undefined,
779291 INFO DEBUG -   schema: undefined,
779292 INFO DEBUG -   table: undefined,
779292 INFO DEBUG -   column: undefined,
779293 INFO DEBUG -   dataType: undefined,
779293 INFO DEBUG -   constraint: undefined,
779293 INFO DEBUG -   file: 'auth.c',
779294 INFO DEBUG -   line: '331',
779294 INFO DEBUG -   routine: 'auth_failed'
779294 INFO DEBUG - }
779295 INFO DEBUG - undefined
779295 INFO DEBUG - /var/www/calckey/packages/backend:
779295 INFO DEBUG -  ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  backend@ migrate: `typeorm migration:run -d ormconfig.js`
779296 INFO DEBUG - Exit status 1
779296 INFO DEBUG -  ELIFECYCLE  Command failed with exit code 1.
779297 INFO DEBUG - + ynh_exit_properly
```

I don't understand what we are supposed to do during the restore process :(
https://github.com/lapineige/calckey_ynh/blob/master/scripts/restore#L84

lapineige commented:
> > Hey @ThatOneCalculator the sha256sum of the archive changed, do you know why ?
> 
> Of which archive? We did reupload our docker image due to an issue.

The tar.gz.

ThatOneCalculator commented:
Yeah we redid the release for docker

lapineige commented:
And that's the file used ?

Ok then nothing to worry about 🙂

lapineige commented:
https://ci-apps-dev.yunohost.org/ci/job/5457
So finally upgrade is working. At least on paper. I will try on my side 🤞

I don't understand how restoration fails. @ThatOneCalculator could you guide me here please ? 🙏

lapineige commented:
> So finally upgrade is working. At least on paper. I will try on my side crossed_fingers

Upgrade fails :
```
2023-02-03 12:52:30,261: DEBUG - + sudo -u calckey env PATH=/opt/node_n/n/versions/node/19/bin:/opt/node_n/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/root/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin pnpm clean-all
2023-02-03 12:52:33,031: DEBUG - 
2023-02-03 12:52:33,032: DEBUG - > calckey@13.1.0 clean-all /var/www/calckey
2023-02-03 12:52:33,032: DEBUG - > pnpm node ./scripts/clean-all.js
2023-02-03 12:52:33,032: DEBUG - 
2023-02-03 12:52:34,496: DEBUG - node:internal/modules/cjs/loader:1056
2023-02-03 12:52:34,496: DEBUG -   throw err;
2023-02-03 12:52:34,496: DEBUG -   ^
2023-02-03 12:52:34,497: DEBUG - 
2023-02-03 12:52:34,497: DEBUG - Error: Cannot find module 'execa'
2023-02-03 12:52:34,497: DEBUG - Require stack:
2023-02-03 12:52:34,497: DEBUG - - /var/www/calckey/scripts/clean-all.js
2023-02-03 12:52:34,497: DEBUG -     at Module._resolveFilename (node:internal/modules/cjs/loader:1053:15)
2023-02-03 12:52:34,497: DEBUG -     at Module._load (node:internal/modules/cjs/loader:898:27)
2023-02-03 12:52:34,497: DEBUG -     at Module.require (node:internal/modules/cjs/loader:1120:19)
2023-02-03 12:52:34,497: DEBUG -     at require (node:internal/modules/helpers:112:18)
2023-02-03 12:52:34,497: DEBUG -     at Object.<anonymous> (/var/www/calckey/scripts/clean-all.js:2:15)
2023-02-03 12:52:34,498: DEBUG -     at Module._compile (node:internal/modules/cjs/loader:1239:14)
2023-02-03 12:52:34,498: DEBUG -     at Module._extensions..js (node:internal/modules/cjs/loader:1293:10)
2023-02-03 12:52:34,498: DEBUG -     at Module.load (node:internal/modules/cjs/loader:1096:32)
2023-02-03 12:52:34,498: DEBUG -     at Module._load (node:internal/modules/cjs/loader:935:12)
2023-02-03 12:52:34,498: DEBUG -     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:84:12) {
2023-02-03 12:52:34,498: DEBUG -   code: 'MODULE_NOT_FOUND',
2023-02-03 12:52:34,498: DEBUG -   requireStack: [ '/var/www/calckey/scripts/clean-all.js' ]
2023-02-03 12:52:34,499: DEBUG - }
2023-02-03 12:52:34,506: DEBUG - 
2023-02-03 12:52:34,508: DEBUG - Node.js v19.4.0
2023-02-03 12:52:34,545: DEBUG -  ELIFECYCLE  Command failed with exit code 1.
2023-02-03 12:52:34,582: DEBUG - + ynh_exit_properly
```

lapineige commented:
Also I had an un-signed yarn repository issue before this one, I had to fix it with `curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -`. Just as https://github.com/YunoHost-Apps/calckey_ynh/blob/testing/scripts/install#L77 was doing before...

So it seems it still calls yarn for some reason :thinking:

lapineige committed:
Improve progression status

lapineige commented:
Should we change the service conf too ? How ? :thinking:
It uses [`/opt/node_n/n/versions/node/19/bin/`]`npm start` right now. Should I hardcode it to `/opt/node_n/n/versions/node/19.4.0/bin/pnpm` ?
We can't let Yunohost manage that path as the help doesn't handle it (right ? ping @yalh76), so we have to manually select the pnpm that `corepack enable` configure ? (ping @ThatOneCalculator if you have any clue)

lapineige committed:
Switch to pnpm

lapineige commented:
!testme

lapineige commented:
Why is there a new CI from a different domain ? :thinking: (and it's not working btw)

ThatOneCalculator commented:
@lapineige corepack enable should automatically select pnpm, if need be you can also add `corepack prepare pnpm@latest --activate`

lapineige committed:
v13.1.1

Hotfix: https://codeberg.org/calckey/calckey/releases/tag/v13.1.1

lapineige committed:
Merge branch 'testing' into master

ThatOneCalculator commented:
`manifest.json` should be updated to 13.1.1 :)

lapineige committed:
It needs the full pnpm path

lapineige commented:
What with the sudden change in the master branch ? :thinking:
@oufmilo I don't understand what you did to the master branch and the way it affected mine ? :thinking:
(also the source url for v13.1.0 was incomplete :)
Also I don't understand why you started to work on the upgrade https://github.com/YunoHost-Apps/calckey_ynh/commit/05eca3cf78257c818cf56d8a1b17f4cb615d4544 elsewhere :thinking:

lapineige committed:
v13.1.1

lapineige commented:
!testme

lapineige commented:
> `manifest.json` should be updated to 13.1.1 :)

Indeed, I was on it but first trying to figure out why there were so many sudden changes that was difficult to understand (github UI/UX being quite bad :/) :sweat_smile:

lapineige committed:
Change pnpm path

lapineige commented:
!testme

lapineige commented:
It can't find pnpm.
I don't know what's the good way to fix this… is that pnpm located locally thanks to corepack ?

ThatOneCalculator commented on scripts/upgrade:
```diff
@@ -135,12 +135,11 @@ ynh_add_nginx_config
 
 pushd "$final_path"
-    ynh_use_nodejs
-    corepack enable
     corepack prepare pnpm@latest --activate
```
`corepack enable` needs to be run before prepare

ThatOneCalculator commented on conf/systemd.service:
```diff
@@ -9,7 +9,7 @@ Group=__APP__
 Environment="__YNH_NODE_LOAD_PATH__"
 Environment="NODE_ENV=production"
-ExecStart=__YNH_NPM__ start
+ExecStart="/opt/node_n/n/versions/node/19/bin/pnpm start"
```
Any reason for hard coding the path instead of letting corepack handle it?

lapineige replied:
How corepack will handle it from the service ?
If I run the service in the folder, will it use the local pnpm (right now it doesn't find it, is it because corepack enable wasn't used ?)

ThatOneCalculator commented on scripts/install:
```diff
@@ -128,14 +128,14 @@ ynh_add_config --template="../conf/default.yml" --destination="$final_path/.conf
 	corepack prepare pnpm@latest --activate
-	ynh_exec_warn_less ynh_exec_as $app env $ynh_node_load_PATH NODE_ENV=production pnpm i
-	ynh_exec_warn_less ynh_exec_as $app env $ynh_node_load_PATH NODE_ENV=production pnpm run build
+	ynh_exec_warn_less ynh_exec_as $app env $ynh_node_load_PATH NODE_ENV=production pnpm install
```
Env doesn't need to be prod for installing modules, but doesn't matter too much either

lapineige replied:
Ok so I will remove it, no need for useless extra code (and complexity for new contributors)

oufmilo committed:
Update systemd.service

oufmilo committed:
Update upgrade

lapineige commented:
!testme

lapineige commented:
[![Test Badge](https://img.shields.io/endpoint?url=https://yunorunner.travelpandas.eu/ci/api/job/9/badge)](https://yunorunner.travelpandas.eu/ci/job/9)
[![](https://yunorunner.travelpandas.eu/ci/summary/9.png)](https://yunorunner.travelpandas.eu/ci/job/9) thanks @oufmilo

lapineige commented:
Ok that's much better, but restoration still fails with the same error :(
https://github.com/YunoHost-Apps/calckey_ynh/pull/12#issuecomment-1415813315

oufmilo committed:
Update restore

lapineige commented:
Thanks @oufmilo. And now we have : 
> sudo -u calckey env PATH=/opt/node_n/n/versions/node/19/bin:/opt/node_n/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin pnpm init
223561 INFO DEBUG -  ERR_PNPM_PACKAGE_JSON_EXISTS  package.json already exists
(https://yunorunner.travelpandas.eu/ci/job/9)

oufmilo committed:
Update restore

ThatOneCalculator commented:
Why was `pnpm init` ever being run?? That's for making a new project 😆

lapineige commented:
Well we don't understand at all what these command do… I'm not into NodeJS, and even less in pnpm…

lapineige committed:
Merge branch 'testing' into master

lapineige commented:
So we should not use pnpm init in any restore or upgrade ?
(even if the app was totally uninstalled before, as restoration can be done after uninstalling ?)

What about the build command ?

lapineige commented:
Ok restore is fixed according to CI, well done @oufmilo !

oufmilo commented:
!testme

lapineige commented:
Upgrade is working too ! (on paper)

🎉

ThatOneCalculator commented:
> So we should not use pnpm init in any restore or upgrade ?
> (even if the app was totally uninstalled before, as restoration can be done after uninstalling ?)
> 
> What about the build command ?

Only build, no init

lapineige commented:
```
248940 DEBUG + sudo -u calckey env PATH=/opt/node_n/n/versions/node/19/bin:/opt/node_n/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/root/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin pnpm clean-all
251739 DEBUG 
251740 DEBUG > calckey@13.1.1 clean-all /var/www/calckey
251740 DEBUG > pnpm node ./scripts/clean-all.js
251740 DEBUG 
252979 DEBUG node:internal/modules/cjs/loader:1063
252979 DEBUG   throw err;
252980 DEBUG   ^
252980 DEBUG 
252980 DEBUG Error: Cannot find module 'execa'
252980 DEBUG Require stack:
252980 DEBUG - /var/www/calckey/scripts/clean-all.js
252980 DEBUG     at Module._resolveFilename (node:internal/modules/cjs/loader:1060:15)
252980 DEBUG     at Module._load (node:internal/modules/cjs/loader:905:27)
252980 DEBUG     at Module.require (node:internal/modules/cjs/loader:1127:19)
252985 DEBUG     at require (node:internal/modules/helpers:112:18)
252985 DEBUG     at Object.<anonymous> (/var/www/calckey/scripts/clean-all.js:2:15)
252985 DEBUG     at Module._compile (node:internal/modules/cjs/loader:1246:14)
252985 DEBUG     at Module._extensions..js (node:internal/modules/cjs/loader:1300:10)
252986 DEBUG     at Module.load (node:internal/modules/cjs/loader:1103:32)
252986 DEBUG     at Module._load (node:internal/modules/cjs/loader:942:12)
252986 DEBUG     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12) {
252986 DEBUG   code: 'MODULE_NOT_FOUND',
252986 DEBUG   requireStack: [ '/var/www/calckey/scripts/clean-all.js' ]
252986 DEBUG }
252986 DEBUG 
252986 DEBUG Node.js v19.6.0
253016 DEBUG  ELIFECYCLE  Command failed with exit code 1.
```

Upgrade failed on my side.
It seems I have missing (node ?) modules.
But I already have an issue since the restoration, where I don't have all modules I guess, and this results in a 502 bad gateway issue.

@ThatOneCalculator do you know how I can reinstall all the (missing) module please ?

edit : fixed with `sudo -u calckey NODE_ENV=production yarn install` (I don't know what I'm doing, I just tried ^^).
Now I'll try to ugprade again.

lapineige committed:
build command is back

According to https://github.com/YunoHost-Apps/calckey_ynh/pull/12#issuecomment-1416866696

lapineige commented:
!testme

ThatOneCalculator commented:
> NODE_ENV=production yarn install

Please DONT do that! We're now using pnpm, not yarn. The correct command is `pnpm install`

lapineige commented:
I wasn't clear : I fixed my current version that was broken, it was Calckey 13.0.5. It was broken because of some node_modules files removed by mistake I believe.
Then I could upgrade to 13.1.1 and it worked like a charm ! 🚀

lapineige commented:
So… CI is perfectly happy, manual upgrade went well…

@oufmilo let's merge it to testing (so it's available to more people) and then wait for more manual test just in case ? Does it work on your side now ?

oufmilo commented:
Yes it's working for me ! 🚀

lapineige commented:
> edit : fixed with `sudo -u calckey NODE_ENV=production yarn install` (I don't know what I'm doing, I just tried ^^).
> Now I'll try to ugprade again.

Turns out when it fails for the same reason on v13.1.2, the same command (with `pnpm install now`) fix it too :)

In that case the issue was triggered by the server running out of space... for some reason some modules disappeared...
